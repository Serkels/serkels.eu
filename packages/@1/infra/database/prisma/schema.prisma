//

//
//
//

//
//
//

datasource db {
  provider = "postgres"
  url      = env("DATABASE_URL")
  // provider = "sqlite"
  // url      = "file:./dev.db"
}

generator client {
  provider = "prisma-client-js"
}

//
//
//

//
//
//

model SignupPayload {
  role    ProfileRole
  name    String
  email   String              @unique
  tokens  VerificationToken[]
  context Json
}

model Profile {
  id        String      @id @default(cuid())
  bio       String?
  bookmarks Bookmark[]
  contacts  Profile[]   @relation("Contacts")
  following Profile[]   @relation("ProfileFollows")
  image     String
  name      String
  role      ProfileRole
  user      User        @relation(fields: [user_id], references: [id], onDelete: Cascade)
  user_id   String      @unique @map("userId")
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  partner                Partner?
  studient               Studient?
  written_messages       Message[] @relation("WrittenMessages")
  followed_by            Profile[] @relation("ProfileFollows")
  participants_in_thread Thread[]  @relation("ParticipantsInThread")
  in_contact_with        Profile[] @relation("Contacts")
}

model Studient {
  id             String     @id @default(cuid())
  citizenship    String?
  city           String?
  field_of_study String?
  interest       Category[] @relation("CategorizedStudientInterest")
  profile        Profile    @relation(fields: [profile_id], references: [id], onDelete: Cascade)
  profile_id     String     @unique
  university     String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  asked_questions      Question[]       @relation("AskedQuestions")
  exchange_threads     ExchangeThread[]
  inbox_threads        InboxThread[]
  participants_in_deal Deal[]
  proposed_exchanges   Exchange[]       @relation("ProposedExchanges")
}

model Partner {
  id            String        @id @default(cuid())
  city          String?
  link          String
  opportunities Opportunity[]
  profile       Profile       @relation(fields: [profile_id], references: [id], onDelete: Cascade)
  profile_id    String        @unique
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
}

//

enum ProfileRole {
  ADMIN
  PARTNER
  STUDIENT
}

//
//
//

//
//
//

model Bookmark {
  id             String       @id @default(cuid())
  exchange       Exchange?    @relation("BookmarkedExchanges", fields: [exchange_id], references: [id])
  exchange_id    String?
  opportunity    Opportunity? @relation("BookmarkedOpportunities", fields: [opportunity_id], references: [id])
  opportunity_id String?
  owner          Profile      @relation(fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id       String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  @@unique(fields: [owner_id, exchange_id], name: "bookmarked_exchange")
  @@unique(fields: [owner_id, opportunity_id], name: "bookmarked_opportunity")
}

//
//
//

//
//
//

model Category {
  id      String           @id @default(cuid())
  context CategoryContext? @default(EXCHANGE)
  rank    Int?
  name    String
  slug    String
  //

  categorized_exchanges         Exchange[]    @relation("CategorizedExchanges")
  categorized_opportunities     Opportunity[] @relation("CategorizedOpportunities")
  categorized_questions         Question[]    @relation("CategorizedQuestions")
  categorized_return_exchanges  Exchange[]    @relation("CategorizedReturnExchanges")
  categorized_studient_interest Studient[]    @relation("CategorizedStudientInterest")
  //

  @@unique(fields: [slug, context], name: "slug_in_context")
}

enum CategoryContext {
  EXCHANGE
  OPPORTUNITY
  FORUM
}

//
//
//

//
//
//

model Exchange {
  id          String       @id @default(cuid())
  category    Category     @relation("CategorizedExchanges", fields: [category_id], references: [id])
  category_id String
  deals       Deal[]
  description String
  expiry_date DateTime     @default(now())
  is_active   Boolean
  is_online   Boolean
  location    String?
  owner       Studient     @relation("ProposedExchanges", fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id    String
  places      Int
  return      Category?    @relation("CategorizedReturnExchanges", fields: [return_id], references: [id])
  return_id   String?
  title       String
  type        ExchangeType
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  bookmarked_exchanges Bookmark[] @relation("BookmarkedExchanges")
}

enum ExchangeType {
  PROPOSAL
  RESEARCH
}

model Deal {
  id             String               @id @default(cuid())
  status         ExchangeThreadStatus @default(IDLE)
  parent         Exchange             @relation(fields: [parent_id], references: [id], onDelete: Cascade)
  parent_id      String
  participant    Studient             @relation(fields: [participant_id], references: [id], onDelete: Cascade)
  participant_id String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  exchange_threads ExchangeThread[]
  //

  @@unique(name: "participant_per_exchange", fields: [parent_id, participant_id])
}

model ExchangeThread {
  id             String   @id @default(cuid())
  last_seen_date DateTime @default(now())
  owner          Studient @relation(fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id       String
  deal           Deal     @relation(fields: [deal_id], references: [id], onDelete: Cascade)
  deal_id        String
  thread         Thread   @relation(fields: [thread_id], references: [id], onDelete: Cascade)
  thread_id      String

  @@unique([owner_id, deal_id])
}

enum ExchangeThreadStatus {
  IDLE
  DENIED
  APPROVED_BY_THE_ORGANIZER
  APPROVED
}

//
//
//

//
//
//

model Opportunity {
  id          String   @id @default(cuid())
  category    Category @relation("CategorizedOpportunities", fields: [category_id], references: [id])
  category_id String
  cover       String
  description String
  expiry_date DateTime
  link        String
  location    String?
  owner       Partner  @relation(fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id    String
  slug        String   @unique
  title       String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  bookmarked_opportunities Bookmark[] @relation("BookmarkedOpportunities")
}

//
//
//

//
//
//

model Question {
  id          String   @id @default(cuid())
  category    Category @relation("CategorizedQuestions", fields: [category_id], references: [id])
  category_id String
  owner       Studient @relation("AskedQuestions", fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id    String
  title       String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
}

//
//
//

//
//
//

model InboxThread {
  id             String   @id @default(cuid())
  is_archived    Boolean  @default(false)
  last_seen_date DateTime @default(now())
  owner          Studient @relation(fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id       String
  thread         Thread   @relation(fields: [thread_id], references: [id], onDelete: Cascade)
  thread_id      String

  @@unique([owner_id, thread_id])
}

model Thread {
  id           String    @id @default(cuid())
  messages     Message[]
  participants Profile[] @relation("ParticipantsInThread")
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  inbox_threads    InboxThread[]
  exchange_threads ExchangeThread[]
}

model Message {
  id        String  @id @default(cuid())
  author    Profile @relation("WrittenMessages", fields: [author_id], references: [id], onDelete: Cascade)
  author_id String
  content   String
  thread    Thread  @relation(fields: [thread_id], references: [id])
  thread_id String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())

  @@unique(name: "unique_date_in_thread", fields: [thread_id, created_at])
}

//
//
//

//
//
//

model PasswordlessToken {
  id         String    @id @default(uuid())
  body       String    @unique
  email      String
  login_date DateTime?
  is_active  Boolean   @default(true)
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
}

//

model Account {
  id                String  @id @default(cuid())
  userId            String
  type              String
  provider          String
  providerAccountId String
  refresh_token     String? @db.Text
  access_token      String? @db.Text
  expires_at        Int?
  token_type        String?
  scope             String?
  id_token          String? @db.Text
  session_state     String?

  user User @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@unique([provider, providerAccountId])
}

model Session {
  id           String   @id @default(cuid())
  sessionToken String   @unique
  userId       String
  expires      DateTime
  user         User     @relation(fields: [userId], references: [id], onDelete: Cascade)
}

model User {
  id            String    @id @default(cuid())
  name          String?
  email         String?   @unique
  emailVerified DateTime?
  image         String?
  accounts      Account[]
  sessions      Session[]

  //

  profile Profile?
}

model VerificationToken {
  identifier String
  token      String   @unique
  expires    DateTime

  payload       SignupPayload? @relation(fields: [payload_email], references: [email], onDelete: Cascade)
  payload_email String?

  @@unique([identifier, token])
}
