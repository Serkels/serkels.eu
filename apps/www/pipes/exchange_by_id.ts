//

import { ID_Schema } from "@1.modules/core/domain";
import { z } from "zod";

//

export const Params_Schema = z.object({ exchange_id: ID_Schema });

export type Params = z.TypeOf<typeof Params_Schema>;
