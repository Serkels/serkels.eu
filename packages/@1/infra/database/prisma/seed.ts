/**
 * Adds seed data to your db
 *
 * @link https://www.prisma.io/docs/guides/database/seed-database
 */
import { faker } from "@faker-js/faker";
import {
  CategoryContext,
  ExchangeThreadStatus,
  ExchangeType,
  Prisma,
  ProfileRole,
  type Deal,
  type Exchange,
} from "@prisma/client";
import { isAfter } from "date-fns";
import dedent from "dedent";
import process from "node:process";
import slugify from "slugify";
import prisma from "../index";

//
//#region 🌱 Seed
//

async function main() {
  await prisma.$transaction([
    prisma.user.deleteMany(),
    //
    prisma.account.deleteMany(),
    prisma.bookmark.deleteMany(),
    prisma.category.deleteMany(),
    prisma.deal.deleteMany(),
    prisma.exchange.deleteMany(),
    prisma.exchangeThread.deleteMany(),
    prisma.inboxMessageNotification.deleteMany(),
    prisma.inboxThread.deleteMany(),
    prisma.message.deleteMany(),
    prisma.notification.deleteMany(),
    prisma.opportunity.deleteMany(),
    prisma.partner.deleteMany(),
    prisma.profile.deleteMany(),
    prisma.question.deleteMany(),
    prisma.signupPayload.deleteMany(),
    prisma.student.deleteMany(),
    prisma.thread.deleteMany(),
    //
  ]);
  console.log("🗑️");

  //

  await categories();
  console.log("🌱 . Categories.");

  console.log("🌱 . Yopmail Students.");
  await student_yopmail();
  await students();
  console.log("🌱 . Students.");
  await partners();
  console.log("🌱 . Partners.");

  await partners_yopmail();
  await partners_vip();
  console.log("🌱 . Partners VIP.");

  await students_bookmarks();
  console.log("🌱 . Students bookmarks.");
  await students_participants_in_exchanges();
  console.log("🌱 . Students participe to exchanges.");

  await students_messages();
  console.log("🌱 . Students sent some messages.");

  await students_awnsers();
  console.log("🌱 . Students awnser to questions.");
}

main()
  .catch((e) => {
    console.error(e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

//
//#endregion
//

//
//#region 🎓 Students
//

async function student_yopmail() {
  await student({ email: "douglas@yopmail.com", name: "Douglas" });
  await student({ email: "johan@yopmail.com", name: "Johan" });
  await student({ email: "dino@yopmail.com", name: "Dino" });
}

async function student({
  email: studient_email,
  name: studient_name,
}: {
  email: string;
  name: string;
}) {
  const { image, name } = {
    image: faker.image.avatar(),
    name: studient_name,
  };

  const [category_autres] = await Promise.all([
    prisma.category.findFirstOrThrow({
      where: { slug: "autres" },
    }),
  ]);

  const exchange_categories_id = (
    await prisma.category.findMany({
      where: { context: CategoryContext.EXCHANGE },
    })
  ).map(({ id }) => id);

  const forum_categories_id = (
    await prisma.category.findMany({
      where: { context: CategoryContext.FORUM },
    })
  ).map(({ id }) => id);

  const updated_at = faker.date.past();
  const asked_questions: Prisma.QuestionCreateNestedManyWithoutOwnerInput = {
    createMany: {
      data: faker.helpers.multiple<Prisma.QuestionUncheckedCreateWithoutOwnerInput>(
        () => ({
          category_id: faker.helpers.arrayElement([
            category_autres.id,
            ...forum_categories_id,
          ]),
          title: faker.company.catchPhrase(),
          created_at: faker.helpers.arrayElement([
            updated_at,
            faker.date.recent({ refDate: updated_at }),
          ]),
          updated_at,
        }),
        { count: { min: 0, max: 5 } },
      ),
    },
  };

  return prisma.student.create({
    data: {
      language: faker.company.name(),
      city: faker.location.city(),
      field_of_study: faker.word.noun(),
      interest: {
        connect: faker.helpers.arrayElements(
          exchange_categories_id.map((id) => ({ id })),
        ),
      },
      university: faker.company.name(),
      //
      asked_questions,
      proposed_exchanges: {
        createMany: {
          data: faker.helpers.multiple(
            () => ({
              is_active: true,
              category_id: faker.helpers.arrayElement([
                category_autres.id,
                ...exchange_categories_id,
              ]),
              created_at: faker.date.past(),
              updated_at: faker.date.recent({ days: 66 }),
              description: faker.lorem.paragraph(),
              is_online: faker.number.int(100) > 50,
              location: faker.location.city(),
              places: faker.number.int({ min: 1, max: 9 }),
              return_id:
                faker.helpers.maybe(() =>
                  faker.helpers.arrayElement([
                    category_autres.id,
                    ...exchange_categories_id,
                  ]),
                ) ?? null,
              title: faker.company.catchPhrase(),
              type: faker.helpers.arrayElement([
                ExchangeType.PROPOSAL,
                ExchangeType.RESEARCH,
              ]),
              expiry_date: faker.helpers.weightedArrayElement([
                { value: null, weight: 1 },
                { value: faker.date.future({ years: 0.1 }), weight: 1 },
                { value: faker.date.recent({ days: 1 }), weight: 2 },
              ]),
            }),
            { count: { min: 0, max: 5 } },
          ),
        },
      },
      profile: {
        create: {
          bio: dedent`
          # ${faker.company.buzzPhrase()}

          > ${faker.company.catchPhrase()}

          ${faker.lorem.paragraphs(faker.number.int({ max: 5 }))},

          ## ${faker.lorem.words()}

          ${faker.lorem.paragraphs()}

          - ${faker.lorem.sentence()}
          - ${faker.lorem.sentence()}
          - ${faker.lorem.sentence()}

          ${faker.lorem.paragraphs()}

          ---

          👋
          `,
          image,
          name,
          role: ProfileRole.STUDENT,
          user: {
            create: {
              email: studient_email,
              image,
              name,
            },
          },
        },
      },
    },
  });
}

async function students() {
  await Promise.all(
    Array.from({ length: 22 }).map(() =>
      student({
        email: faker.internet.email().toLowerCase(),
        name: faker.person.fullName(),
      }),
    ),
  );
}

//
//#endregion
//

//
//#region 🤝 Partners
//

async function partner({ email, name }: { email: string; name: string }) {
  const image = faker.image.avatar();

  const opportunity_categories_id = (
    await prisma.category.findMany({
      where: {
        OR: [{ context: CategoryContext.OPPORTUNITY }, { context: null }],
      },
    })
  ).map(({ id }) => id);

  return prisma.partner.create({
    data: {
      profile: {
        create: {
          bio: dedent`
          # ${faker.company.buzzPhrase()}

          > ${faker.company.catchPhrase()}

          ${faker.lorem.paragraphs(faker.number.int({ max: 5 }))},
          `,
          image,
          name,
          role: ProfileRole.PARTNER,
          user: {
            create: {
              email,
              image,
              name,
            },
          },
        },
      },
      link: faker.internet.url(),
      city: faker.location.city(),
      opportunities: {
        createMany: {
          data: faker.helpers.multiple(
            () => ({
              category_id: faker.helpers.arrayElement(
                opportunity_categories_id,
              ),
              cover: faker.image.url(),
              description: faker.lorem.paragraphs({ max: 9, min: 2 }),
              link: faker.internet.url(),
              location: faker.location.city(),
              slug: slugify(faker.lorem.sentence()).slice(0, 16).toLowerCase(),
              title: faker.lorem.sentence(),
              expiry_date: faker.date.future(),
            }),
            { count: { min: 1, max: 5 } },
          ),
        },
      },
    },
  });
}

async function partners() {
  return await Promise.all(
    faker.helpers.multiple(
      () =>
        partner({
          email: faker.internet.email().toLowerCase(),
          name: faker.person.fullName(),
        }),
      { count: { max: 15, min: 10 } },
    ),
  );
}

async function partners_yopmail() {
  return prisma.partner.create({
    data: {
      profile: {
        create: {
          bio: dedent`
          # One Does University

          > One does not simply implement, it teaches.
          `,
          image: "https://avatars.githubusercontent.com/u/9634140?s=200&v=4",
          name: "One Does University",
          role: ProfileRole.PARTNER,
          user: {
            create: {
              email: "onedoesuniversity@yopmail.com",
              image:
                "https://avatars.githubusercontent.com/u/9634140?s=200&v=4",
              name: "One Does University",
            },
          },
        },
      },
      link: faker.internet.url(),
      city: "Paris",
      opportunities: {
        createMany: {
          data: {
            category_id: await prisma.category
              .findFirstOrThrow({
                select: { id: true },
                where: {
                  name: "Cours de langues",
                  context: CategoryContext.OPPORTUNITY,
                },
              })
              .then(({ id }) => id),
            cover: faker.image.url(),
            description: dedent`
            ## Cours de JavaScript

            > JavaScript est un langage de programmation créé en 1995 par Brendan Eich.
            > Il est le premier langage de programmation à être implémenté par le navigateur web.

            Parce que JavaScript est sur utilisé dans le monde du web, il est toujours utilie pour créer des applications web.
            `,
            link: faker.internet.url(),
            location: "Paris",
            slug:
              slugify("Cours de JavaScript").slice(0, 16).toLowerCase() +
              "-" +
              faker.string.uuid(),
            title: "Cours de JavaScript",
            expiry_date: faker.date.future(),
          },
        },
      },
    },
  });
}

async function partners_vip() {
  await partner({
    email: "ahmadali.fr@gmail.com",
    name: "Ahmad Ali University",
  });
  await partner({
    email: "juliet.meere.uee@gmail.com",
    name: "Juliet Meere University",
  });
}

//
//#endregion
//

//
//#region 📦 Categories
//

interface Category {
  name: string;
  context: CategoryContext | null;
  rank?: number;
  slug?: string;
}

async function create_categories(categories: Category[]) {
  for (let rank = 0; rank < categories.length; rank++) {
    const category = categories[rank];
    if (!category) continue;
    const { name, context } = category;
    const slug = category.slug ?? slugify(name).toLocaleLowerCase();

    await prisma.category.create({
      data: {
        name,
        slug,
        context,
        rank: category.rank === undefined ? rank + 1 : category.rank,
      },
    });
  }
}

async function categories() {
  await create_categories(
    [
      {
        name: "Autres",
        rank: 0,
      },
    ].map((draft) => ({ ...draft, context: null })),
  );

  //

  await create_categories(
    [
      { name: "Activités​" },
      { name: "Cours de langues" },
      { name: "Notes de cours" },
      { name: "Soutien académique​" },
    ]
      .reverse()
      .map((draft) => ({ ...draft, context: CategoryContext.EXCHANGE })),
  );
  //

  await create_categories(
    [
      { name: "Cours et Matières" },
      { name: "Logement" },
      { name: "Stages et Alternance" },
      { name: "Santé et Bien-être" },
      { name: "Vie étudiante" },
      { name: "Transport et Mobilité" },
      { name: "Technologie et Informatique" },
      { name: "International et Échanges" },
      { name: "Finances et Bourses" },
    ]
      .reverse()
      .map((draft) => ({ ...draft, context: CategoryContext.FORUM })),
  );

  //

  await create_categories(
    [
      { name: "Cours de langues" },
      { name: "Soutien académique" },
      { name: "Logements" },
      { name: "Job étudiant" },
      { name: "Concours" },
      { name: "Aides financières" },
      { name: "Activités socio-culturelles​" },
      { name: "Vie associative" },
    ]
      .reverse()
      .map((draft) => ({ ...draft, context: CategoryContext.OPPORTUNITY })),
  );
}

//
//#endregion
//

//
//#region 🔖 Bookmarks
//

async function students_bookmarks() {
  const students = await prisma.student.findMany({
    include: { profile: true },
  });
  const opportunities = await prisma.opportunity.findMany();
  const exchanges = await prisma.exchange.findMany();

  await prisma.bookmark.createMany({
    data: [
      ...Array.from({ length: 20 }).map(() => ({
        owner_id: faker.helpers.arrayElement(students).profile_id,
        opportunity_id: faker.helpers.arrayElement(opportunities).id,
      })),
      ...Array.from({ length: 20 }).map(() => ({
        owner_id: faker.helpers.arrayElement(students).profile_id,
        exchange_id: faker.helpers.arrayElement(exchanges).id,
      })),
    ],
    skipDuplicates: true,
  });
}

//
//#endregion
//

//
//#region 🔄 Exchanges
//

async function students_participants_in_exchanges() {
  const students = await prisma.student.findMany();
  const exchanges = await prisma.exchange.findMany({
    include: { owner: true },
  });

  for (const exchange of exchanges) {
    const other_students = students.filter(
      ({ id }) => id !== exchange.owner_id,
    );
    const exchange_id = exchange.id;

    await Promise.all(
      faker.helpers
        .arrayElements(other_students, { min: 0, max: exchange.places })
        .map(async (participant) => {
          const participant_id = participant.id;

          const { id: deal_id } = await prisma.deal.upsert({
            create: {
              parent_id: exchange_id,
              participant_id,
              status: faker.helpers.arrayElement(
                Object.values(ExchangeThreadStatus),
              ),
              created_at: faker.date.past(),
              updated_at: faker.date.recent({ days: 66 }),
            },
            update: {
              updated_at: faker.date.recent({ days: 33 }),
            },
            where: {
              participant_per_exchange: {
                parent_id: exchange_id,
                participant_id,
              },
            },
          });

          const { thread_id } = await prisma.exchangeThread.create({
            data: {
              deal: { connect: { id: deal_id } },
              owner: { connect: { id: exchange.owner_id } },
              thread: {
                create: {
                  created_at: faker.date.past(),
                  updated_at: faker.date.past(),
                  participants: {
                    connect: [
                      { id: exchange.owner.profile_id },
                      { id: participant.profile_id },
                    ],
                  },
                  messages: {
                    createMany: {
                      data: faker.helpers.multiple(
                        () => ({
                          author_id: faker.helpers.arrayElement([
                            { id: exchange.owner.profile_id },
                            { id: participant.profile_id },
                          ]).id,
                          content: faker.lorem.sentences(),
                          created_at: faker.helpers.weightedArrayElement([
                            { value: faker.date.past(), weight: 1 },
                            {
                              value: faker.date.recent({ days: 66 }),
                              weight: 5,
                            },
                          ]),
                        }),
                        { count: { min: 1, max: 5 } },
                      ),
                    },
                  },
                },
              },
            },
          });

          await prisma.exchangeThread.create({
            data: { deal_id, thread_id, owner_id: participant_id },
          });

          const updated_exchange = await prisma.exchange.findFirstOrThrow({
            include: {
              deals: { where: { status: ExchangeThreadStatus.APPROVED } },
            },
            where: { id: exchange_id },
          });

          const is_active = is_active_exchange(updated_exchange);
          await prisma.exchange.update({
            data: { is_active },
            where: { id: exchange_id },
          });
        }),
    );
  }
}

export function is_active_exchange(
  exchange: Pick<Exchange, "expiry_date" | "places"> & { deals: Deal[] },
) {
  return exchange.deals.length < exchange.places;
}

//
//#endregion
//

//
//#region 📝 Questions/Awnsers
//

async function students_awnsers() {
  const students = await prisma.student.findMany();
  const questions = await prisma.question.findMany({
    include: { owner: { select: { id: true } } },
  });

  for (const question of questions) {
    const { owner } = question;
    const other_students = faker.helpers.arrayElements(
      students.filter(({ id }) => id !== owner.id),
    );

    const { answers } = await prisma.question.update({
      data: {
        answers: {
          createMany: {
            data: faker.helpers
              .arrayElements(other_students, {
                min: 0,
                max: other_students.length,
              })
              .map<Prisma.AnswerCreateManyParentInput>(({ id: owner_id }) => ({
                content: faker.company.catchPhrase(),
                owner_id,
              })),
          },
        },
      },
      include: { answers: true },
      where: { id: question.id },
    });

    if (faker.helpers.rangeToNumber({ min: 0, max: 100 }) > 0) {
      const accepted_answer = faker.helpers.arrayElement([
        { id: null },
        ...answers,
      ]);
      await prisma.question.update({
        data: { accepted_answer_id: accepted_answer.id },
        where: { id: question.id },
      });
    }
    // await faker.helpers.maybe(
    //   () =>
    //     prisma.question.update({
    //       data: { accepted_answer_id: faker.helpers.arrayElement(q_answer).id },
    //       where: { id: question.id },
    //     }),
    //   { probability: 0.2 },
    // );
  }
}

//
//#endregion
//

//
//#region 📬 Messages
//

async function students_messages() {
  const students = await prisma.student.findMany();

  for (const student of students) {
    const other_students = faker.helpers.arrayElements(
      students.filter(({ id }) => id !== student.id),
    );

    const contacts = faker.helpers.arrayElements(other_students, {
      min: 0,
      max: 10,
    });

    await prisma.profile.update({
      data: {
        contacts: {
          connect: contacts.map(({ profile_id }) => ({ id: profile_id })),
        },
        following: {
          connect: faker.helpers
            .arrayElements(contacts)
            .map(({ profile_id }) => ({ id: profile_id })),
        },
      },
      where: { id: student.profile_id },
    });

    await Promise.all(
      contacts.map(async (recipient_student) => {
        const last_seen_date = faker.date.recent();
        const thread_created_at = faker.date.past({
          refDate: last_seen_date,
        });
        const thread_updated_at = faker.date.between({
          from: thread_created_at,
          to: last_seen_date,
        });

        if (
          await prisma.inboxThread.findFirst({
            where: {
              owner_id: recipient_student.id,
              thread: {
                participants: {
                  some: { id: student.profile_id },
                },
              },
            },
          })
        )
          return;

        const thread = await prisma.thread.create({
          data: {
            created_at: thread_created_at,
            updated_at: thread_updated_at,
            participants: {
              connect: [
                { id: student.profile_id },
                { id: recipient_student.profile_id },
              ],
            },
            inbox_threads: {
              createMany: {
                data: [
                  {
                    owner_id: student.id,
                    last_seen_date: last_seen_date,
                  },
                  {
                    owner_id: recipient_student.id,
                    last_seen_date: faker.date.between({
                      from: thread_created_at,
                      to: last_seen_date,
                    }),
                  },
                ],
              },
            },
            messages: {
              createMany: {
                data: faker.helpers.multiple(
                  () => {
                    const message_created_at = faker.date.between({
                      from: thread_created_at,
                      to: thread_updated_at,
                    });
                    return {
                      author_id: faker.helpers.arrayElement([
                        { id: student.profile_id },
                        { id: recipient_student.profile_id },
                      ]).id,
                      content: faker.lorem.sentences(),
                      created_at: message_created_at,
                      updated_at: message_created_at,
                    };
                  },
                  { count: { min: 1, max: 11 } },
                ),
              },
            },
          },
          select: {
            id: true,
            updated_at: true,
            inbox_threads: {
              select: { last_seen_date: true },
              where: { owner_id: recipient_student.id },
            },
            messages: {
              select: {
                id: true,
                created_at: true,
              },
              orderBy: { created_at: "desc" },
              take: 1,
              where: { author: { id: student.profile_id } },
            },
          },
        });

        const recipient_inbox = thread.inbox_threads.at(0);
        const message = thread.messages.at(0);
        if (!recipient_inbox) {
          return;
        }
        if (!message) {
          return;
        }
        {
          const { last_seen_date } = recipient_inbox;
          const { created_at } = message;
          await prisma.inboxMessageNotification.create({
            data: {
              message: { connect: { id: message.id } },
              notification: {
                create: {
                  type: "INBOX_NEW_MESSAGE",
                  created_at,
                  read_at: isAfter(last_seen_date, created_at)
                    ? last_seen_date
                    : null,
                  owner_id: recipient_student.profile_id,
                },
              },
            },
          });
        }
      }),
    );
  }
}

//
//#endregion
//
