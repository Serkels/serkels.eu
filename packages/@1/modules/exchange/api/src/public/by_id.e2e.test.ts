//

import {
  create_douglas_student,
  create_joedart_student,
  create_johan_student,
  create_masterclass_category,
  create_slap_exchange,
} from "@1.infra/database/seeding";
import prisma, { empty_database, migrate } from "@1.infra/database/testing";
import {
  create_nextauth_header,
  createCallerFactory,
  router,
} from "@1.modules/trpc";
import {
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  setSystemTime,
  test,
} from "bun:test";
import exchange_by_id from "./by_id";

//

const NEXTAUTH_SECRET = "ðŸ”‘";

beforeAll(() => {
  process.env["NEXTAUTH_SECRET"] = NEXTAUTH_SECRET;
});

//

beforeAll(empty_database);
beforeAll(migrate);
beforeAll(() => {
  setSystemTime(new Date("2011-11-11"));
});

//

describe("visitor", () => {
  test("should return a sanitize exchange", async () => {
    const caller = createCallerFactory(router({ by_id: exchange_by_id }));
    const trpc = caller({
      prisma,
    } as any);
    const exchange = await trpc.by_id("slap_exchange_id");
    expect(exchange).toMatchSnapshot();
  });
});

describe("connected studient", () => {
  let nextauth_header: Awaited<ReturnType<typeof create_nextauth_header>>;

  beforeAll(async () => {
    nextauth_header = await create_nextauth_header({
      secret: "ðŸ”‘",
      token: {
        from: import.meta.url,
        profile: { id: "douglas_profile_id", role: "STUDENT" },
      },
    });
  });

  test("should return a sanitize exchange", async () => {
    const caller = createCallerFactory(router({ by_id: exchange_by_id }));
    const trpc = caller({
      headers: { ...nextauth_header },
      prisma,
    } as any);
    const exchange = await trpc.by_id("slap_exchange_id");
    expect(exchange).toMatchSnapshot();
  });
});

//

beforeEach(async function seed() {
  await create_masterclass_category(prisma);
  const douglas_student_id = await create_douglas_student(prisma);
  const johan_student_id = await create_johan_student(prisma);
  const joedart_student = await create_joedart_student(prisma);
  const { id: exchange_id } = await create_slap_exchange(
    prisma,
    joedart_student,
  );
  await prisma.exchange.update({
    data: {
      deals: {
        createMany: {
          data: [
            {
              id: "slap_exchange_id_douglas_student_id_deal_id",
              participant_id: douglas_student_id,
              status: "APPROVED",
            },
            {
              id: "slap_exchange_id_johan_student_id_deal_id",
              participant_id: johan_student_id,
              status: "APPROVED",
            },
          ],
        },
      },
    },
    where: { id: exchange_id },
  });
});

afterEach(async function empty_delete_entries() {
  await prisma.category.deleteMany({});
  await prisma.exchange.deleteMany({});
  await prisma.user.deleteMany({});
});
