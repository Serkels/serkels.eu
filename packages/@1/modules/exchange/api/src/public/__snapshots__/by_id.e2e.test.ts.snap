// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`visitor should return a sanitize exchange 1`] = `
{
  "category": {
    "context": "EXCHANGE",
    "id": "masterclass_category_id",
    "name": "Masterclass",
    "rank": null,
    "slug": "masterclass",
  },
  "category_id": "masterclass_category_id",
  "created_at": 2011-11-11T00:00:00.000Z,
  "deals": [
    {
      "id": "slap_exchange_id_douglas_student_id_deal_id",
      "participant": undefined,
      "status": "APPROVED",
    },
    {
      "id": "slap_exchange_id_johan_student_id_deal_id",
      "participant": undefined,
      "status": "APPROVED",
    },
  ],
  "description": "https://vulf.co/courses/dart/lectures/45940163",
  "expiry_date": null,
  "id": "slap_exchange_id",
  "is_active": true,
  "is_online": true,
  "location": null,
  "owner": {
    "profile": {
      "id": "joedart_profile_id",
      "image": "https://picsum.photos/200/300",
      "name": "Joe Dart",
    },
    "university": "Université du Michigan",
  },
  "owner_id": "joedart_student_id",
  "places": 1,
  "return": null,
  "return_id": null,
  "title": "Slap",
  "type": "PROPOSAL",
  "updated_at": 2011-11-11T00:00:00.000Z,
}
`;

exports[`connected studient should return a sanitize exchange 1`] = `
{
  "category": {
    "context": "EXCHANGE",
    "id": "masterclass_category_id",
    "name": "Masterclass",
    "rank": null,
    "slug": "masterclass",
  },
  "category_id": "masterclass_category_id",
  "created_at": 2011-11-11T00:00:00.000Z,
  "deals": [
    {
      "id": "slap_exchange_id_douglas_student_id_deal_id",
      "participant": {
        "profile": {
          "id": "douglas_profile_id",
        },
      },
      "status": "APPROVED",
    },
    {
      "id": "slap_exchange_id_johan_student_id_deal_id",
      "participant": undefined,
      "status": "APPROVED",
    },
  ],
  "description": "https://vulf.co/courses/dart/lectures/45940163",
  "expiry_date": null,
  "id": "slap_exchange_id",
  "is_active": true,
  "is_online": true,
  "location": null,
  "owner": {
    "profile": {
      "id": "joedart_profile_id",
      "image": "https://picsum.photos/200/300",
      "name": "Joe Dart",
    },
    "university": "Université du Michigan",
  },
  "owner_id": "joedart_student_id",
  "places": 1,
  "return": null,
  "return_id": null,
  "title": "Slap",
  "type": "PROPOSAL",
  "updated_at": 2011-11-11T00:00:00.000Z,
}
`;
